### PR TITLE
Add logStores method and various other improvements

### PR DESCRIPTION
**Pull request by rhgills**

### CODE REVIEW

**Comment by dfed** [comment]
LGTM. Though let's figure out that build issue.

**Comment by dfed** [comment]
Rebased you off of master. You were missing your .travis.yml file. Merge when green

**Comment by rhgills** [comment]
Made some more changes in:
https://github.com/square/Aardvark/commit/dca365f436372e0b7d8478d7a65b7d17e171060f
and 
https://github.com/square/Aardvark/commit/d153981e9a92834e158bd79de6d38bdc6819abaa.

In dca365f436372e0b7d8478d7a65b7d17e171060f:
- Adding the default bug reporting gesture was creating a defaultLogStore, instead of using the dispatch_once'd one, which had the new properties. No reason for that, as it set no special configuration. Changed to request the defaultLogStore from the distributor instead.

- Tests were failing because the `defaultLogStore` implementation was calling through to itself. In addition, from the dispatch_once docs:

"The result of using a predicate with automatic or dynamic storage (including Objective-C instance variables) is undefined."

I replaced `dispatch_once` using an ivar with a `BOOL` to determine if this is the first time the accessor is called. If the accessor is called for the first time, then we'll lazily instantiate, as before.

In d153981e9a92834e158bd79de6d38bdc6819abaa:
Made the ARKLogTableViewController only set its title if it doesn't already have one. Want to be able to set more specific titles than "Logs" for when we're using multiple ARKLogTableViewControllers in our app.

Let me know if still approved!

**Comment by dfed** [comment]
Love the work with the `dispatch_once`... that was on my todo list to fix. However it needs some thread-safety improvements before it can be merged.

File: Log Viewing/ARKLogTableViewController.m [hunk 1]
```diff
@@ -109,16 +109,17 @@ - (void)viewWillDisappear:(BOOL)animated;
     [super viewWillDisappear:animated];
 }
 
-- (void)viewDidLoad;
-{
+- (void)viewDidLoad; {
```
Comment (dfed) [inline Log Viewing/ARKLogTableViewController.m]: non-blocking sylte nit: newline after `;`

File: Log Viewing/ARKLogTableViewController.m [hunk 2]
```diff
@@ -109,16 +109,17 @@ - (void)viewWillDisappear:(BOOL)animated;
...
-    
+
     self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
     self.tableView.rowHeight = 34.0;
     self.tableView.dataSource = self;
     self.tableView.delegate = self;
-    
-    self.title = NSLocalizedString(@"Logs", @"Title of log viewing screen.");
+
+    if (!self.title) {
```
Comment (dfed) [inline Log Viewing/ARKLogTableViewController.m]: `self.title.length == 0`. Unless you specifically only want to change this if the title is nil, at which point I'd prefer a `!= nil`

File: Logging/ARKLogDistributor.h [hunk 3]
```diff
@@ -38,6 +38,9 @@
 /// Convenience method to store a reference to the default log store. Lazily creates a log store when accessed for the first time if one is not already set.
 @property (atomic, strong, readwrite) ARKLogStore *defaultLogStore;
 
+/// Returns all instances of `ARKLogStore` that are currently registered as observers on this log distributor.
+@property (atomic, strong, readonly) NSSet *logStores;
```
Comment (dfed) [inline Logging/ARKLogDistributor.h]: `(copy, readonly)`

**Comment by dfed** [comment]
Oh, and please update the version number in the pod spec to 1.0.2

**Comment by rhgills** [comment]
All feedback should be addressed. Please take another look!

File: AardvarkTests/ARKLogDistributorTests.m [hunk 4]
```diff
@@ -143,7 +143,13 @@ - (void)test_setLogMessageClass_appendedLogsAreCorrectClass;
...
 {
     ARKLogDistributor *logDistributor = [ARKLogDistributor new];
-    XCTAssertNotNil(logDistributor.defaultLogStore);
+    ARKLogStore *defaultLogStore = logDistributor.defaultLogStore;
+    XCTAssertNotNil(defaultLogStore);
+    XCTAssertEqualObjects(defaultLogStore.name, @"Default");
+    XCTAssertFalse(defaultLogStore.prefixNameWhenPrintingToConsole);
+
+    // Should return the same instance on subsequent property accesses.
+    XCTAssertEqual(logDistributor.defaultLogStore, defaultLogStore);
```
Comment (dfed) [inline AardvarkTests/ARKLogDistributorTests.m]: nice

File: Logging/ARKLogDistributor.m [hunk 5]
```diff
@@ -34,9 +34,13 @@ @interface ARKLogDistributor ()
 
 @property (atomic, assign) Class internalLogMessageClass;
 @property (atomic, weak) ARKLogStore *weakDefaultLogStore;
+@property (atomic, strong, readonly) NSRecursiveLock *defaultLogStorePropertyLock;
```
Comment (dfed) [inline Logging/ARKLogDistributor.m]: non-blocking style nit: don't need the `strong` (#5 removes all the excessive modifiers)

File: Logging/ARKLogDistributor.m [hunk 6]
```diff
@@ -80,6 +83,9 @@ - (instancetype)init;
 #endif
     
     _logObservers = [NSMutableArray new];
+
+    _defaultLogStorePropertyLock = [NSRecursiveLock new];
```
Comment (dfed) [inline Logging/ARKLogDistributor.m]: Why a recursive lock? You aren't expecting reentrancy are you?

File: Logging/ARKLogDistributor.m [hunk 7]
```diff
@@ -91,28 +97,43 @@ - (instancetype)init;
...
 
 - (void)setDefaultLogStore:(ARKLogStore *)logStore;
 {
-    // Remove the old log store.
-    [self removeLogObserver:self.weakDefaultLogStore];
-    
-    if (logStore) {
-        // Add the new log store. The logObserver array will hold onto the log store strongly.
-        [self addLogObserver:logStore];
+    [self.defaultLogStorePropertyLock lock];
```
Comment (dfed) [inline Logging/ARKLogDistributor.m]: Technically this lock only needs to be around setting the `weakDefaultLogStore`, but I kinda dig having it around the whole process

File: Logging/ARKLogDistributor.m [hunk 8]
```diff
@@ -91,28 +97,43 @@ - (instancetype)init;
...
+    * are mututally exclusive.
+    */
+    [self.defaultLogStorePropertyLock lock];
+    {
+        if (!self.defaultLogStoreAccessorCalled && self.weakDefaultLogStore == nil) {
+            // Lazily create a default log store if none exists.
+            ARKLogStore *defaultLogStore = [[ARKLogStore alloc] initWithPersistedLogFileName:[NSStringFromClass([self class]) stringByAppendingString:@"_DefaultLogStore"]];
+            defaultLogStore.name = @"Default";
+            defaultLogStore.prefixNameWhenPrintingToConsole = NO;
+            self.defaultLogStore = defaultLogStore;
```
Comment (dfed) [inline Logging/ARKLogDistributor.m]: ah, here's the reentrancy

**Comment by dfed** [comment]
Looks like there's only the non-blocking style nit. LGTM. Merge when green!

File: Logging/ARKLogDistributor.m (hunk 7 above)
Comment (rhgills) [inline Logging/ARKLogDistributor.m]: :+1:

**Comment by rhgills** [comment]
merging! thanks!

**Pull request merged**